blogposts/2022/introducting-migrator-service.md

```diff
@@ -0,0 +1,148 @@
+2. We respond to a _reduction of pods_ due to a dirty database, but do not pay particular attention to the original error condition. In fact, any original error (SQL, network, timeout, or deadlock) may have often been obscured by a unrelated bug that occurs when two error libraries we use interact in a certain way: [cockroachdb/errors](https://github.com/cockroachdb/errors) and [hashicorp/go-multierror](https://github.com/hashicorp/go-multierror). (We're currently investigating this issue as well, so stay tuned for another blog post!)
+3. We deploy to Cloud by starting new services that immediately attempt to run migrations.
+
+The first observation serves as loose evidence of a correlation between concurrent index creation and dirty database issues. This correlation does, albeit unscientifically, match reality.
+
+The second observation implies that we never actually check the _error message_ that occurs when the database was first marked as dirty. Looking back, a number of downtime incidents we originally attributed to migrations not completing within our Kubernetes health check timeout are also explainable by this deadlock condition. We've been increasingly relying on concurrent index creation, and _any_ contention of a migration advisory lock will cause the deadlock to occur. And, by the third observation, we have plenty of lock contention when multiple pods start up at the same time looking to update the database.
+
+With a new theory in hand, we can finally attack the problem in a focused manner.
+
+We changed the migration process to gracefully deals with concurrent index creation by dropping the advisory lock and instead coordinating with concurrent migration attempts within the Postgres catalog itself. Of particular interest is the `pg_stat_progress_create_index` system view, which [reports index creation progress](https://www.postgresql.org/docs/12/progress-reporting.html). The migrator is also intelligence enough to detect invalid indexes are repair them rather than flagging a human to hold its hand.
+
+After another round of testing, migrations finally behave according to intuition.
+
+## Conclusion
+
+We've long since outgrown the way in which we run database migrations. We've known for a while that there _was_ a problem, and felt the pain along with our users for an embarrassingly long time. After digging a bit deeper, better understand the problem space, and breaking a bit of our ossified code and deployment process along the way, I hope we're delivering a quality of life improvement to Sourcegraphers and our users.
+
+If your next upgrade does **not** crash, please tweet me at [@ericfritz](https://twitter.com/ericfritz).
+
+### Related resources
+
+We've updated several documentation pages which are worth a look for existing site administrators:
+
+- [How to run `migrator` operations](https://docs.sourcegraph.com/admin/how-to/manual_database_migrations)
```

Comment:
@DaedalusG Any other docs you'd like to call out in-particular?

Reply:
re. https://github.com/sourcegraph/about/pull/5070#discussion_r808309950

Reply:
Hmm actually that would be the main one - perhaps this doc
https://docs.sourcegraph.com/admin/how-to/rollback_database

I mainly made this comment after running into a few folks who were confused about the term migratory and assumed it was related to database migrations between databases

[https://sourcegraph.slack.com/archives/C02BJ8T258D/p1644968878319539](https://sourcegraph.slack.com/archives/C02BJ8T258D/p1644968878319539)
[https://sourcegraph.slack.com/archives/C01JR51JR5J/p1645032270381909](https://sourcegraph.slack.com/archives/C01JR51JR5J/p1645032270381909)